core/src/avm2/globals/flash/net/XMLSocket.as

```diff
@@ -1,23 +1,85 @@
-package flash.net {
-    import flash.events.EventDispatcher;
-    import __ruffle__.stub_constructor;
+package flash.net
+{
     import __ruffle__.stub_method;
+    import flash.net.Socket;
+    import flash.events.ProgressEvent;
+    import flash.events.DataEvent;
+    import flash.utils.ByteArray;
 
-    public class XMLSocket extends EventDispatcher {
-        public function XMLSocket(host: String = null, port: int = 0) {
-            stub_constructor("flash.net.XMLSocket");
+    // NOTE: Extending from Socket is a implementation detail. It is not possible for SWFs to
+    // use Socket methods as Flash's playerglobal doesn't include those methods, unless
+    // SWF author intentionally tried, which wouldn't make much sense.
+    public class XMLSocket extends Socket
+    {
+        // Events ioError, connect, close, securityError are inherited from Socket.
+        // connected, timeout are inherited from Socket.
+
+        private var tempBuf:ByteArray = new ByteArray();
+
+        public function XMLSocket(host:String = null, port:int = 0)
+        {
+            super(null, 0);
+
+            this.addEventListener(ProgressEvent.SOCKET_DATA, this.socketDataListener);
+
+            if (host != null)
+            {
+                this.connect(host, port);
+            }
         }
 
-        public function close(): void {
-            stub_method("flash.net.XMLSocket", "close");
+        override public function connect(host:String, port:int):void
+        {
+            if (host == null)
+            {
+                // FIXME: We should take the movie "host" and connect to it here.
+                stub_method("flash.net.XMLSocket", "connect() with null host");
```

Comment:
I think the implementation of `LocalConnection.domain` looks up the movie domain- you should be able to reuse the same logic here.